### ISSUE TITLE
Orientation in speech/voice output if phone is upright

### ISSUE THREAD

**Issue by geocachermgo** [cgeo/cgeo#7152]
##### Detailed steps causing the problem:
if i hold the phone horizontally I get the correct distance and directions like "200m 4 Uhr".

##### Actual behavior after performing these steps:
If the phone is upright I get arbitrary directions.

##### Expected behavior after performing these steps:
The phone should recognize that the phone is upright and switch from y-axis to z axis as "12 Uhr"
(x-axis being left/right if you look at the screen, y-axis being up-down and z front-back)

##### Version of c:geo used:
2018.09.17

##### Is the problem reproducible for you?
Yes

##### System information:
```
--- System information ---
Device: GT-I9505 (jfltexx, samsung)
Android version: 5.0.1
Android build: LRX22C.I9505XXUHPK2
c:geo version: 2018.09.17
Google Play services: disabled - 13.2.80 (020308-211705629)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide own/found: true
Map strategy: auto
HW acceleration: enabled (default state)
System language: de_DE
System date format: dd.MM.yyyy
Debug mode active: no
System internal c:geo dir: /data/data/cgeo.geocaching (1,3 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (1,2 GB free) external non-removable
Geocache data: /storage/extSdCard/Android/data/cgeo.geocaching/files/GeocacheData (6,5 GB free) external removable
Database: /storage/emulated/0/Android/data/cgeo.geocaching/files/databases/data (9,8 MB) on user storage
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
geocaching.com: Logged in (Anmeldung OK) / PREMIUM
Geocaching.com date format: MM/dd/yyyy
Installed c:geo plugins: none
--- End of system information ---
```

##### Other comments and remarks:
e.g. Reference to other issues, projects, sources, etc.

**Comment by Lineflyer** [comment]
Programming wise I cannot judge whether this is easily doable, however that is AFAIR the reason why we implemented the ability to disable the use of the magnetic sensor and fallback to GPS direction calculation by using the toogle in the compass view.

**Comment by SammysHP** [comment]
Screen rotation should be respected, so this is a bug.

**Comment by SammysHP** [comment]
So you don't mean portrait or landscape mode (screen orientation), but phone orientation in general. That is a different issue and requires GPS.

**Comment by geocachermgo** [comment]
I don't mean screen orientation (portrait/landscape). With this switching  everything works fine.

I do not suggest to switch from phone orientation to movement orientation but to switch from y-axis to z axis as "12 Uhr if the gravity direction is in the direction of y-axis.
horizontal: 
![waagrecht](https://user-images.githubusercontent.com/43813636/48509676-101b7f80-e852-11e8-985a-3b7383da2873.jpg)
upright:
Right now if the phone is upright (gravity direction is in the direction of y-axis.) the compass gives arbitrary directions which is ok, since the compass is on the screen and cannot show a direction perpendicular to the screen.
![senkrecht](https://user-images.githubusercontent.com/43813636/48509714-1c074180-e852-11e8-9399-9cf74d4510a2.jpg)
The voice will say 6 o clock if its upright e.g. in my shirt pocket.
I want that the evaluation of the direction (now in screen area, x, y ) to the waist area (x,y,)
So if the back of my phone is directed to the target it will say "12 o clock" and if the right edge is directed to the target it will say "3 o clock"

**Comment by geocachermgo** [comment]
@eddiemuc : Are you interested?

**Comment by eddiemuc** [comment]
> @eddiemuc : Are you interested?

Have some issues on my plate right now, but I can put it in my list :-)

**Comment by eddiemuc** [comment]
Hi @geocachermgo , I will take a look at this, but it seems I have some reading/learning to do first to learn how this sensor stuff is handled in Android and how it was previously implemented in c:geo. So this might take a while...

**Comment by geocachermgo** [comment]
The request is from 2018, I'm that someone is interested. I will be patient until 2022 :) and hpoe the the new knowledge will be helpful for you.

**Comment by eddiemuc** [comment]
![image](https://user-images.githubusercontent.com/6909759/91640038-14ab2b00-ea1b-11ea-9ed4-285f84cb0d4a.png)

**Comment by eddiemuc** [comment]
Notes to self: 
tried to use `SensorManager.remapCoordinateSystem(rotateMatrix, AXIS_X, AXIS_Z, adjustedRotationMatrix);` in `MagnetometerAndAccelerometerProvider` and `RotationProvider` in case of an upright device. Looks good in emulator, don't know whether it will work on real device...
`

**Comment by eddiemuc** [comment]
Hi @geocachermgo, I created an alpha-version for this. It is far from ready in the sense of being mergeable, but I was able to produce some promising results with two devices of mine.
Question: could you maybe find the time, install the debug-APK I build for this and see if this would work for you?

Remark: I have added a new toggle button to the compass screen with the settings "device is flat" and "device is upright" (currently hardcoded). "device is flat" should give you the same behaviour as before, while "device is upright" should give you the expected behaviour. At least that's the idea.

You will find the debug APK here: 
* Direct link: https://ci.cgeo.org/job/cgeo%20pull%20request/3478/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk
* CI result page: https://ci.cgeo.org/job/cgeo%20pull%20request/3478/   

Thanks!

**Comment by eddiemuc** [comment]
* Direct link to APK: https://ci.cgeo.org/job/cgeo%20pull%20request/3480/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk
* CI page with APK: https://ci.cgeo.org/job/cgeo%20pull%20request/3480/

**Comment by geocachermgo** [comment]
First try: did not install
![IMG_20200829_191305](https://user-images.githubusercontent.com/43813636/91642503-25b06800-ea2c-11ea-86f9-c824c190dcc9.jpg)
I will now completly deinstall cgeo and try again.

**Comment by geocachermgo** [comment]
Works fine, including voice output.
For me the manual switch would also be fine instead of the automatic switch by direction of the gravity.
Thank you very much!

**Comment by geocachermgo** [comment]
Works not only for cache but aslo for waypoint in real drive-by test.
Did not update distance and bearing at the field test, when screen was locked and black. This could be a power-save coordinates accuracy  issue. Back at home there was a voice output even with locked screen.

**Comment by eddiemuc** [comment]
Thanks a lot for feedback! I will try to find a solution for finding out phone position (flat vs upright) automatically and implement the feature in a product-ready way.

**Comment by geocachermgo** [comment]
Shall I test some more?
Is anything from the https://square.github.io/leakcanary/ needed?

**Comment by eddiemuc** [comment]
Hi @geocachermgo , If everything goes well I will implement the feature sometime today, maybe in next couple of hours (but have to do the family laundry first...). If you want you can test the next alpha version afterwards. I will inform you once it is ready.

I did ask you yesterday for the alpha test because I was not sure how the Android sensors work and whether I used them correctly. Thanks to your efforts I know now that I did. The rest is normal development. Thanks a lot for you help!

**Comment by geocachermgo** [comment]
There is one leaks entry from cgeo:
┬───
│ GC Root: System class
│
├─ android.provider.FontsContract class
│    Leaking: NO (CgeoApplication↓ is not leaking and a class is never leaking)
│    ↓ static FontsContract.sContext
├─ cgeo.geocaching.CgeoApplication instance
│    Leaking: NO (Application is a singleton)
│    ↓ CgeoApplication.mLoadedApk
│                      ~~~~~~~~~~
├─ android.app.LoadedApk instance
│    Leaking: UNKNOWN
│    ↓ LoadedApk.mServices
│                ~~~~~~~~~
├─ android.util.ArrayMap instance
│    Leaking: UNKNOWN
│    ↓ ArrayMap.mArray
│               ~~~~~~
├─ java.lang.Object[] array
│    Leaking: UNKNOWN
│    ↓ Object[].[1]
│               ~~~
├─ android.util.ArrayMap instance
│    Leaking: UNKNOWN
│    ↓ ArrayMap.mArray
│               ~~~~~~
├─ java.lang.Object[] array
│    Leaking: UNKNOWN
│    ↓ Object[].[0]
│               ~~~
├─ cgeo.geocaching.maps.routing.BRouterServiceConnection instance
│    Leaking: UNKNOWN
│    ↓ BRouterServiceConnection.onServiceConnectedCallback
│                               ~~~~~~~~~~~~~~~~~~~~~~~~~~
├─ cgeo.geocaching.maps.mapsforge.v6.-$$Lambda$NewMap$1JK6pPzIiNuIN7hxFK_6gQyteYs instance
│    Leaking: UNKNOWN
│    ↓ -$$Lambda$NewMap$1JK6pPzIiNuIN7hxFK_6gQyteYs.f$0
│                                                   ~~~
╰→ cgeo.geocaching.maps.mapsforge.v6.NewMap instance
​     Leaking: YES (ObjectWatcher was watching this because cgeo.geocaching.maps.mapsforge.v6.NewMap received Activity#onDestroy() callback and Activity#mDestroyed is true)
​     key = de97e16b-6c81-443f-a57c-33c7f0c8e6ba
​     watchDurationMillis = 5177
​     retainedDurationMillis = 176

METADATA

Build.VERSION.SDK_INT: 29
Build.MANUFACTURER: Xiaomi
LeakCanary version: 2.4
App process name: cgeo.geocaching
Analysis duration: 15794 ms

**Comment by eddiemuc** [comment]
HI @geocachermgo, the LeakCanary tool is installed automatically if you install the alpha `debug` APK-versions of c:geo from CI server (where I was pointing to yesterday). During development it is running all the time and indeed it gives me messages and warnings since I started developing for c:geo and even before I changed a single line of code. Unfortunately I don't yet know how to interpret these.

The good thing is: I am almost sure these messages are not related to the change I did for the "orientation in speech/voice" issue.

@moving-bits how do we handle LeadCanary warnings in general?

**Comment by moving-bits** [comment]
Depends on how we can make sense of them. Sometimes it's as easy as using a `WeakReference` instead of the actual variable for eg. `activity` or `context`, but sometimes... - I'm no expert at reading/interpreting the LeakCanary info :-(

**Comment by eddiemuc** [comment]
Hi @geocachermgo , I created a PR for a possible production-ready version of this fix. It adds the following line into the compass view (marked with a red box):
![image](https://user-images.githubusercontent.com/6909759/91662624-dcbbea80-eae3-11ea-979c-b522e8ba4031.png)
On the left you see the estimated heading of the device (behind it the estimated device position: upright or lying). On the right you see the raw orientation sensor values (azimuth, pitch, roll) as returned by the phone sensors (device heading and orientation is based on these values).

I assume this PR will not make it to the next release since beta test is already running. So it has to wait until the release after that. If you are interested in testing this upfront (with debug version again...) you find the APK here:
* Direct link: https://ci.cgeo.org/job/cgeo%20pull%20request/3496/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk 
* CI build page: https://ci.cgeo.org/job/cgeo%20pull%20request/3496/

This topic is rather fiddly because there are so many variables (device movements, portrait/landscape view, turning compass to GPS-only and such). I could not test all possible combinations but did my best. If you find problems, don't hesitate to tell me.

**Comment by geocachermgo** [comment]
Who cares about release version? :-) Will it appear in the next nightly? Then I will test it there. Today is very bad weather the whole day.
Did you include damping/averaging (for e.g. using.it on a bike, I do not know if its neccessary)?

**Comment by eddiemuc** [comment]
> Who cares about release version? :-) Will it appear in the next nightly? Then I will test it there. Today is very bad weather the whole day.

Unfortunately it will only appear in nightly once it was merged to master, and then it would also go into next release which afaik is already closed. I don't have permission to perform such a merge myself (usually done by @moving-bits ).

> Did you include damping/averaging (for e.g. using.it on a bike, I do not know if its neccessary)? 

Yes, that would have been a good idea. Alas, I did not think of that. It might be that switch is already "damped" because the sensors used for orientation react slowly, but I did not verify this. If necessary I will add damping.

**Comment by Lineflyer** [comment]
> Unfortunately it will only appear in nightly once it was merged to master, and then it would also go into next release which afaik is already closed. I don't have permission to perform such a merge myself (usually done by @moving-bits ).

Yes, let us please wait until branched before we merge this change. This allows more time before next release to test it.
@geocachermgo Please be patient for some days. Once you see its merged it will appear in the next nightly.

**Comment by geocachermgo** [comment]
I will test it separatly until Wednesday next week. Real bike test depending on weather.

**Comment by geocachermgo** [comment]
First test result looks good.
I could trigger the switching several times per second. Damping (via EWMA averaging oder switch delay (0.8s)) might be usefull. But I did not get a wrong voice output up to now.

**Comment by geocachermgo** [comment]
First bug: After switching to the browser and back the device heading and (azimuth, pitch, roll) were frozen. Then I switched to androsensor (orientation looks fine), back to c:geo heading is moving again.

**Comment by geocachermgo** [comment]
I like the digital display of the device heading. Its more precise readable and seems to have less "friction" than the compass rose.
Could you add the target direction?
Are the azimuth pitch roll information a debug feature or planned to be released?
If, then they should get a text, like "orientation".
And a small improvment: use %3d so that the [Upright] text is not jumping around if we are at 0 / 359°

**Comment by geocachermgo** [comment]
Regarding friction: Device alligned with the cache (247°), but the rose is not in the device direction:
(No problem of the new feature) 
![IMG_20200830_233554](https://user-images.githubusercontent.com/43813636/91670054-24f5ff80-eb1a-11ea-8392-d0918049be46.jpg)

**Comment by eddiemuc** [comment]
Hi @geocachermgo , 
thanks for the already great feedback. I will look into all the bugs you reported and also add damping according to your suggestions (might take me a few days though)

> Are the azimuth pitch roll information a debug feature or planned to be released?

If it's up to me then yes, but I would value the groups opinion on this too.

**Comment by geocachermgo** [comment]
Suggestion: add one fraction digit (decimal place, Nachkommastelle) to the heading?
At my device this seems to be possible (display of value without fraction is quite stable).
Is is possible to decide by the noise if and how many fraction digits are meaningfull?

**Comment by eddiemuc** [comment]
@geocachermgo 
What would you think of the idea to add possibility to select "device orientation" with options "Auto-Detect", "Lying" and "Upright" ("AUto-detect" would use the currently implemented algo plus damping plus whatever bugs we find, while the other two would set the orientation fixed).

I just don't know yet how to put this into the GUI in a good way

**Comment by geocachermgo** [comment]
Suggestion for GUI : there is already an entry in main menue, "System" regarding orientation sensor, I would add the option there. And additionally: tap on "Upright" or "Lying" to get a selection menu.

**Comment by geocachermgo** [comment]
Bug1 (Frozen): did not happen again
Bug 2: Google Map , Satellit are not working

**Comment by Lineflyer** [comment]
> Bug 2: Google Map , Satellit are not working

AFAICS you tried a debug build. These do not contain any keys for Google maps, thus its not working in these versions.

**Comment by eddiemuc** [comment]
> > Bug 2: Google Map , Satellit are not working
> 
> AFAICS you tried a debug build. These do not contain any keys for Google maps, thus its not working in these versions.

Good thing you wrote that, I would probably have searched hours for this one... :-)

**Comment by geocachermgo** [comment]
Some 10s after locked screen position is not updated any more (probably not related to App change but to energy management of device), direction still changes with rotation of device, distance stays.

**Comment by eddiemuc** [comment]
![image](https://user-images.githubusercontent.com/6909759/91881349-a87e3080-ec81-11ea-80f3-2ed9c8d5e015.png)

**Comment by eddiemuc** [comment]
Hi @geocachermgo , I created a new debug build trying to fix some (but unfortunately not all) bugs you found:
* adding an explicit three-way-button where you can select "auto", "lying" and "upright"
* for auto: adding damping (1 second) so value does not change too often
* fixed layout for heading as well as azimuth/pitch/roll (was not so easy... `format("% 3.1f")` does not work because the font is not fixed-sized) and adding labels

I did not check yet on your issues regarding frozen heading and locked screen position after 10s -> this is more complicated for me to reproduce (esp in emulator)

**Comment by eddiemuc** [comment]
New debug APK is here: https://ci.cgeo.org/job/cgeo%20pull%20request/3516/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk

**Comment by geocachermgo** [comment]
Technically very good.
UI concept also good.
UI:
1)Bug: if GPS only: Upright/Lying/Auto is changeable but display text (upright, lying, auto) is not changing, only colour bar.
1b) (not serious) shall we call it upright/lying/truth?
2) Suggestion: Change text to Upright/Flat (my choice) or Standing/Lying
3) I fould "Device Heading" better than "Device" May be: "Device: Heading xxx         Orientation xx/yy/zz"
4) I do not get the logic of the colour lines: Why is Lying, Upright orange and auto gray. I would suggest the other way (orange=automatic / more sensors) active
5) minor bug (fix not needed?) if I'm in upright (auto) mode, switch to other app and back: Then the app starts with Lying and switches back to Upright after a second.

**Comment by geocachermgo** [comment]
> I did not check yet on your issues regarding frozen heading and locked screen position after 10s -> this is more complicated for me to reproduce (esp in emulator)
- Freeze never happend again (wild guess: no event from sensors?)
- It took me quite a while to get the Locus Track recording running. Im almost sure that the problem is that the OS takes away the GPS signal from c:geo
I will make a comparison with the Nightly.

**Comment by eddiemuc** [comment]
> 1. Bug: if GPS only: Upright/Lying/Auto is changeable but display text (upright, lying, auto) is not changing, only colour bar.

For GPS-only, algorithm does not work anyway. device orientation stuff will now be removed when this option is active

> 2. Suggestion: Change text to Upright/Flat (my choice) or Standing/Lying

Done

> 3. I liked "Device Heading" better than "Device" May be: "Device: Heading xxx         Orientation xx/yy/zz"?

Done

> 4. I dont get the logic of the colour lines: Why is Lying, Upright orange and auto gray? I would suggest the other way (orange=automatic / more sensors) active

I changed it

> 5. minor bug (fix not needed?) if I'm in upright (auto) mode, switch to other app and back: Then the app starts with Lying and switches back to Upright after a second.

This happens because sensor reading starts anew after such a switch and thus it takes a second until new value is accepted. I changed GUI -> now in this case the state "unknown" is shown for one sec ("unknown" is treated like "flat" underneath).

New debug APK: https://ci.cgeo.org/job/cgeo%20pull%20request/3518/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk

**Comment by geocachermgo** [comment]
I'll give you the nickname "Fast Eddie" (lkarke).
So I have to dive deep into nitpicking, since all the changes seem to work fine (only coarse test during lunch break)
The orientation values can toggle between 0° and -0°. 
If the value is rounded to 0° the sign should be omitted.
If the tests dont reveal anything else I would say it ready for production. Congrats!

**Comment by eddiemuc** [comment]
> I'll give you the nickname "Fast Eddie" (Clarke).

Well you are very fast with testing too, and you don't even have a nightly build available for that. I bet there was never a better tested issue before it even got merged into master :-) Thanks a lot for your effort!

I included a fix for the toggling 0°/-0°-values, here is the latest debug APK:
https://ci.cgeo.org/job/cgeo%20pull%20request/3520/artifact/main/build/outputs/apk/basic/debug/cgeo-debug.apk 

I guess pretty soon we have to start waiting for the release after next release...

**Comment by eddiemuc** [comment]
@geocachermgo PR was just merged to master, so change should appear in nightly builds now. No more need to install debug APKs :-)

**Comment by geocachermgo** [comment]
It's in the nighly and works fine (including the +-0) fix.
Thanks again.
Is there something else to be done where I can help (e.g. feature description in docu)?
Now I have to check how to prevent the OS from killing the backgroung activity to really use the voice featute with the phone in the pocket.

**Comment by geocachermgo** [comment]
@Lineflyer is a manual update needed?

**Comment by eddiemuc** [comment]
HI @geocachermgo , I assume the right place to edit would be the manual here: https://cgeo.droescher.eu/de/compass (or https://cgeo.droescher.eu/en/compass for english version). But I think it is too early for that, since the manual has no "release process" (whatever is edited is visible immidiately for the public). So my guess would be that this is only edited shortly before or after the public release of a feature. @Lineflyer  is that more or less correct?

**Comment by geocachermgo** [comment]
I just found out that the setting is applied not only at the compass mode, but also in the list display. The direction of the small gray white arrow changes between upright an flat mode. I like that!
@eddiemuc, was that by purpose?
![IMG_20200927_132910](https://user-images.githubusercontent.com/43813636/94363900-c34b9580-00c5-11eb-9f48-9e91e9f38db1.jpg)

**Comment by eddiemuc** [comment]
> @eddiemuc, was that by purpose?

No that was not on purpose :-) But I think I know why this is happening (because I store the compass setting globally in the background, and this setting is then used centrally in the plae calculation bearings). If this is a wanted side effect I will do nothing about it. I was not aware that this bearings are also used in other places

**Comment by geocachermgo** [comment]
For me it's a wanted/appreciated effect, but since it's a little bit surprising it should be mentioned in the manual.

**Comment by Lineflyer** [comment]
> because I store the compass setting globally in the background, and this setting is then used centrally in the plae calculation bearings).

Interesting, so it will also influence the blue arrow on map AFAIR, which might be good as well.

**Comment by geocachermgo** [comment]
Yes, it does.

**Comment by eddiemuc** [comment]
Hm do you think then that the setting of device orientation (Auto, flat or upright) should be made a global one? Currently this is only settable through compass and defaults to Auto

**Comment by geocachermgo** [comment]
I personally would prefer if all settings could be done via the settings menu. At the moment this is not the case (eg red circle is only available an the map menu). So if the compass behaviour can only be set in the compass thats consistent and not too far away.
